### ISSUE TITLE
Make sure mex-files are compatible with 64bit Matlab

### ISSUE THREAD

**Issue by JohannesPfeifer** [DynareTeam/dynare#1225]
See http://de.mathworks.com/help/matlab/matlab_external/upgrading-mex-files-to-use-64-bit-api.html

**Comment by MichelJuillard** [comment]
@JohannesPfeifer, do have reasons to believe that it has not been done systematically? We have been using mwSize and mwIndex for very long now.

**Comment by JohannesPfeifer** [comment]
No, I haven't, but it is the first time I have that Matlab boilerplate warning seen appearing. It may be related to 32bit being phased out. So we should check this systematically at some point.

**Comment by JohannesPfeifer** [comment]
@MichelJuillard What we should do is call the `mex`-command with the `largeArrayDims`-flag (which is going to supersede the current default `-compatibleArrayDims`-flag) and then check whether the result are the same. When using
```
if strcmp(mexext,'mexw64')
    eval(['mex -largeArrayDims -O LINKFLAGS="$LINKFLAGS /export:Dynamic" COMPFLAGS="/TP" ' basename '_dynamic.c ' basename '_dynamic_mex.c'])
    eval(['mex -largeArrayDims -O LINKFLAGS="$LINKFLAGS /export:Static" COMPFLAGS="/TP" ' basename '_static.c ' basename '_static_mex.c'])
  else
    eval(['mex -O LINKFLAGS="$LINKFLAGS /export:Dynamic" COMPFLAGS="/TP" ' basename '_dynamic.c ' basename '_dynamic_mex.c'])
    eval(['mex -O LINKFLAGS="$LINKFLAGS /export:Static" COMPFLAGS="/TP" ' basename '_static.c ' basename '_static_mex.c'])
end
``` 
on the `example1_use_dll` everything works fine.

Do we already cross-compile the 64bit-dll using this flag?

**Comment by MichelJuillard** [comment]
@JohannesPfeifer this needs to be investigated further. We don't use the "mex" command to compile the mex files, but directly gcc. For cross-compilation, I suspect that it is the same. I need to check what exactly is triggered by the -largeArrayDims option of the "mex" command

**Comment by JohannesPfeifer** [comment]
@MichelJuillard My example above referred to the `dyn_mex.m`, but I guess the same issue is relevant for the gcc-compilation of the prepackaged mex-files

**Comment by JohannesPfeifer** [comment]
@MichelJuillard Looking at e.g. `mjdgges.c` we have
``` 
void
mexFunction(int nlhs, mxArray *plhs[],
            int nrhs, const mxArray *prhs[])

{
  unsigned int m1, n1, m2, n2;
```
If I understand it correctly, the `int` should be `mwSize`

**Comment by MichelJuillard** [comment]
Yes, it should. So, I will have to go through all mex files. But, I guess that we have still to worry about compatibility with 32bit Matlab

**Comment by JohannesPfeifer** [comment]
@MichelJuillard If I understand Mathworks correctly, compatibility with 32bit is assured via a compiler flag (`-compatibleArrayDims` for 32bit vs `-largeArrayDims`). Moreover, Matlab 9 does not support 32bit systems anymore. Therefore, we could just implement this going forward.

**Comment by JohannesPfeifer** [comment]
Matlab R2017a seems to not support 32bit mex-files anymore (see http://www.dynare.org/phpBB3/viewtopic.php?f=1&t=14705)

Thus, it seems we need to upgrade following the steps at https://de.mathworks.com/help/matlab/matlab_external/upgrading-mex-files-to-use-64-bit-api.html

**Comment by houtanb** [comment]
@JohannesPfeifer I believe your message on July 17 is accurate. The 32-bit API is still available by passing the flag `-compatibleArrayDims`, it's just that mex files are now built against the 64-bit API by default. For more details, see the table under "MEX builds with 64-Bit API by default" on https://www.mathworks.com/help/matlab/release-notes.html

**Comment by JohannesPfeifer** [comment]
@houtanb Yes, but I would prefer to go full 64bit going forward. Also, we need to distinguish two things:
1. The precompiled mex-files we ship with Dynare. It seems with R2017a we (again) lose compatibility with all older Dynare version (as the user reported for 4.4.3). This will require changes in the build-system, I guess.
2. Compilation at run-time when e.g. using `use_dll`. Changes here should be straightforward as outlined above.
